tracecontext/src/main/java/brave/propagation/tracecontext/TracestateFormat.java

```diff
@@ -69,11 +74,14 @@ static boolean isLetterOrNumber(char c) {
     return (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9');
   }
 
-  @Override
-  public boolean onEntry(
+  static boolean isValidTracestateValueChar(char c) {
+    return c >= ' ' && c <= '~' && c != ',' && c != '=';
+  }
+
+  @Override public boolean onEntry(
     Tracestate target, String buffer, int beginKey, int endKey, int beginValue, int endValue) {
     if (!validateKey(buffer, beginKey, endKey)) return false;
-    if (!validateValue(buffer, beginValue, beginValue)) return false;
+    if (!validateValue(buffer, beginValue, endValue)) return false;
     return target.put(buffer.substring(beginKey, endKey), buffer.substring(beginValue, endValue));
```

Comment:
```suggestion
    // TODO: consider if we want to keep a ref of the string instead. it will be cheaper than substring
    return target.put(buffer.substring(beginKey, endKey), buffer.substring(beginValue, endValue));
```